### PR TITLE
feat: add serverCertificateUri in DRM advanced config

### LINKED ISSUE #1906: Change DRM config for serverCertificate to serverCertificateUri
**Have you read the [FAQ](https://bit.ly/ShakaFAQ) and checked for duplicate open issues?**
Yes

**Is your feature request related to a problem? Please describe.**
No

**Describe the solution you'd like**
Instead of the app needing to request the server cert, DrmEngine should be capable of doing it for you on demand.  The serverCertificate config should be changed into serverCertificateUri.

**Describe alternatives you've considered**
None come to mind, other than the status quo.

### PR DESCRIPTION
**Pull request by valotvince**
## Description

Fixes #1906

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to
  not work as expected)
- [ ] This change requires a documentation update

## Checklist:

- [x] I have signed the Google CLA <https://cla.developers.google.com>
- [x] My code follows the style guidelines of this project
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have verified my change on multiple browsers on different platforms
- [x] I have run `./build/all.py` and the build passes
- [x] I have run `./build/test.py` and all tests pass

Works when running alone but doesn't when running with all the other tests (same thing on master)
> Chrome 89.0.4389.128 (Mac OS 10.15.7) Offline stores, plays, and deletes protected content with a persistent license FAILED

### CODE REVIEW

File: lib/util/error.js [hunk 1]
```diff
@@ -817,6 +817,13 @@ shaka.util.Error.Code = {
    */
   'INIT_DATA_TRANSFORM_ERROR': 6016,
 
+  /**
+   * The server certificate request failed.  This could be a timeout, a
+   * network failure, or a rejection by the server.
```
Comment (TheModMaker) [inline lib/util/error.js:822]: I'd suggest removing these examples, it gives the impression these are the only possible failures.  I'd suggest just saying `The network request for the server certificate failed` (and keep the `error.data[0]` part).

File: lib/media/drm_engine.js [hunk 2]
```diff
@@ -2151,6 +2208,10 @@ shaka.media.DrmEngine = class {
       if (advancedConfig.sessionType) {
         drmInfo.sessionType = advancedConfig.sessionType;
       }
+
+      if (advancedConfig.serverCertificateUri) {
```
Comment (TheModMaker) [inline lib/media/drm_engine.js:2212]: This should be `!drmInfo.serverCertificateUri` like line 2204.

File: lib/media/drm_engine.js [hunk 3]
```diff
@@ -1975,7 +2023,8 @@ shaka.media.DrmEngine = class {
 
     // TODO: refactor, don't stick drmInfos onto MediaKeySystemConfiguration
     shaka.media.DrmEngine.processDrmInfos_(
-        config['drmInfos'], licenseServers, serverCerts, initDatas, keyIds);
+        config['drmInfos'], licenseServers, serverCerts,
+        serverCertificateUris, initDatas, keyIds);
 
     if (serverCerts.length > 1) {
```
Comment (TheModMaker) [inline lib/media/drm_engine.js:2029]: You should update this warning for the URIs too.  This should also warn that the buffer takes precedence over the URI.

File: externs/shaka/player.js [hunk 4]
```diff
@@ -545,6 +546,10 @@ shaka.extern.EmsgInfo;
  *   A key-system-specific server certificate used to encrypt license requests.
  *   Its use is optional and is meant as an optimization to avoid a round-trip
  *   to request a certificate.
+ * @property {string} serverCertificateUri
+ *   <i>Defaults to '', e.g., server certificate will be requested from the
+ *   given URI if serverCertificate is NOT provided. Can be filled in by
+ *   advanced DRM config.</i>
```
Comment (TheModMaker) [inline externs/shaka/player.js:552]: This is the advanced DRM config.  Also I'd suggest the following formatting:

```
<i>Defaults to <code>''</code>.</i><br>
If given, will make a request to the given URI to get the server certificate.
This is ignored if <code>serverCertificate</code> is set.
```

File: demo/common/assets.js [hunk 5]
```diff
@@ -369,7 +369,9 @@ shakaAssets.testAssets = [
       .addFeature(shakaAssets.Feature.SUBTITLES)
       .addFeature(shakaAssets.Feature.WEBM)
       .addFeature(shakaAssets.Feature.OFFLINE)
-      .addLicenseServer('com.widevine.alpha', 'https://cwip-shaka-proxy.appspot.com/no_auth'),
+      .addLicenseServer('com.widevine.alpha', 'https://cwip-shaka-proxy.appspot.com/no_auth')
+      .setExtraConfig({drm: {advanced: {'com.widevine.alpha': {serverCertificateUri:
+        'https://storage.googleapis.com/wvmedia/cert/cert_license_widevine_com_uat.bin'}}}}),
```
Comment (TheModMaker) [inline demo/common/assets.js:374]: This isn't really needed as the CDM will request it for us anyway.  I'd suggest dropping the demo changes to make things simpler.

File: lib/media/drm_engine.js [hunk 6]
```diff
@@ -451,26 +452,66 @@ shaka.media.DrmEngine = class {
...
-        if (!supported) {
-          shaka.log.warning('Server certificates are not supported by the ' +
-                            'key system.  The server certificate has been ' +
-                            'ignored.');
-        }
-      } catch (exception) {
-        throw new shaka.util.Error(
+        const response = await this.playerInterface_.netEngine.request(
+            shaka.net.NetworkingEngine.RequestType.SERVER_CERTIFICATE,
+            request).promise;
```
Comment (TheModMaker) [inline lib/media/drm_engine.js:469]: I'd suggest splitting this into two statements to make it easier to read.

```js
const operation = this.playerInterface_.netEngine.request(...);
const response = await operation.promise;
```

File: lib/media/drm_engine.js [hunk 7]
```diff
@@ -451,26 +452,66 @@ shaka.media.DrmEngine = class {
...
+
+        const shakaErr = new shaka.util.Error(
             shaka.util.Error.Severity.CRITICAL,
             shaka.util.Error.Category.DRM,
-            shaka.util.Error.Code.INVALID_SERVER_CERTIFICATE,
-            exception.message);
+            shaka.util.Error.Code.SERVER_CERTIFICATE_REQUEST_FAILED,
+            error);
+
+        this.onError_(shakaErr);
```
Comment (TheModMaker) [inline lib/media/drm_engine.js:484]: Why use `onError_` instead of throwing?

File: test/media/drm_engine_unit.js [hunk 8]
```diff
@@ -972,16 +974,91 @@ function testDrmEngine(useMediaCapabilities) {
...
+      const variants = manifest.variants;
+      await drmEngine.initForPlayback(variants, manifest.offlineSessionIds,
+          useMediaCapabilities);
+
+      expect(fakeNetEngine.request).toHaveBeenCalledWith(
+          shaka.net.NetworkingEngine.RequestType.SERVER_CERTIFICATE,
+          jasmine.objectContaining({
+            uris: [serverCertificateUri],
+            method: 'GET',
+          }));
```
Comment (TheModMaker) [inline test/media/drm_engine_unit.js:1011]: You can use:

```js
fakeNetEngine.expectRequest(
    serverCertificateUri,
    shaka.net.NetworkingEngine.RequestType.SERVER_CERTIFICATE)
```

File: lib/media/drm_engine.js [hunk 9]
```diff
@@ -451,26 +452,66 @@ shaka.media.DrmEngine = class {
...
-        this.currentDrmInfo_ &&
-        this.currentDrmInfo_.serverCertificate &&
-        this.currentDrmInfo_.serverCertificate.length) {
+    if (!this.mediaKeys_ || !this.currentDrmInfo_) {
+      return;
+    }
+
+    if (this.currentDrmInfo_.serverCertificateUri &&
+        (!this.currentDrmInfo_.serverCertificate ||
+        !this.currentDrmInfo_.serverCertificate.length)) {
```
Comment (TheModMaker) [inline lib/media/drm_engine.js:461]: Does the linter complain if you indent this one more space?  Normally the style is to align with the contents of the parenthesis, but our linter isn't that great about continuations.

**Review by TheModMaker — CHANGES_REQUESTED** [review]

File: demo/common/assets.js (hunk 5 above)
Comment (valotvince) [reply demo/common/assets.js:374]: Yep, it was just to have a working example with the serverCertificateUri. I could drop it or comment to explain this shouldn't be needed, but it is here to demonstrate the feature, WYT ?

**Comment by valotvince** [comment]
@TheModMaker 👋 I've pushed some edits base on your review, thanks !

**Pull request merged**